testing/e2e/extensions_test.go

```diff
@@ -0,0 +1,277 @@
+			Type:  "handle",
+			Event: handleEvt,
+			Env: map[string]string{
+				"SENSU_E2E_HANDLE_ERROR":    "i/o timeout",
+				"SENSU_E2E_HANDLE_RESPONSE": "{}",
+			},
+			ExpErr: true,
+		},
+		{
+			Name:  "logic error",
+			Type:  "handle",
+			Event: handleEvt,
+			Env: map[string]string{
+				"SENSU_E2E_HANDLE_RESPONSE": `{"error": "some error"}`,
+			},
+			ExpErr: true,
+		},
+		{
+			Name:  "no error and not filtered",
+			Type:  "filter",
+			Event: filterEvt,
+			Env: map[string]string{
+				"SENSU_E2E_FILTER_RESPONSE": "{}",
+			},
+		},
+		{
+			Name:  "no error and filtered",
+			Type:  "filter",
+			Event: filterEvt,
+			Env: map[string]string{
+				"SENSU_E2E_FILTER_RESPONSE": `{"filtered": true}`,
+			},
+		},
+		{
+			Name:  "RPC error",
+			Type:  "filter",
+			Event: filterEvt,
+			Env: map[string]string{
+				"SENSU_E2E_FILTER_ERROR":    "i/o timeout",
+				"SENSU_E2E_FILTER_RESPONSE": "{}",
+			},
+			ExpErr: true,
+		},
+		{
+			Name:  "logic error",
+			Type:  "filter",
+			Event: filterEvt,
+			Env: map[string]string{
+				"SENSU_E2E_FILTER_RESPONSE": `{"error":"internal error"}`,
+			},
+			ExpErr: true,
+		},
+		{
+			Name:  "no error",
+			Type:  "mutate",
+			Event: mutateEvt,
+			Env: map[string]string{
+				"SENSU_E2E_MUTATE_RESPONSE": `{"mutatedEvent":"{}"}`,
+			},
+		},
+		{
+			Name:  "RPC error",
+			Type:  "mutate",
+			Event: mutateEvt,
+			Env: map[string]string{
+				"SENSU_E2E_MUTATE_ERROR":    "i/o timeout",
+				"SENSU_E2E_MUTATE_RESPONSE": "{}",
+			},
+			ExpErr: true,
+		},
+		{
+			Name:  "logic error",
+			Type:  "mutate",
+			Event: mutateEvt,
+			Env: map[string]string{
+				"SENSU_E2E_MUTATE_RESPONSE": `{"error":"internal error"}`,
+			},
+			ExpErr: true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%s %s", test.Type, test.Name), func(t *testing.T) {
+			cleanup := newExtension(t, ports[0], test.Env)
+
+			body, _ := json.Marshal(test.Event)
+			resp, err := client.R().SetBody(body).Put(
+				fmt.Sprintf("/events/%s/%s", test.Event.Entity.ID, test.Event.Check.Name))
+			if err != nil {
+				t.Fatal(err)
+			}
+			if status := resp.StatusCode(); status >= 400 {
+				t.Fatalf("bad status: %d", status)
+			}
+
+			out, err := cleanup()
+			if err != nil && !test.ExpErr {
+				fmt.Println(out)
+				t.Fatal(err)
+			}
+
+			switch test.Type {
+			case "handle":
+				if err == nil {
+					assert.Equal(t, test.Env["SENSU_E2E_HANDLE_RESPONSE"], out)
+				} else {
+					assert.Equal(t, test.Env["SENSU_E2E_HANDLE_ERROR"], out)
+				}
+			case "filter":
+				if err == nil {
+					assert.Equal(t, test.Env["SENSU_E2E_FILTER_RESPONSE"], out)
+				} else {
+					assert.Equal(t, test.Env["SENSU_E2E_FILTER_ERROR"], out)
+				}
+			case "mutate":
+				if err == nil {
+					assert.Equal(t, test.Env["SENSU_E2E_MUTATE_RESPONSE"], out)
+				} else {
+					assert.Equal(t, test.Env["SENSU_E2E_MUTATE_ERROR"], out)
+				}
+			default:
+				t.Fatalf("invalid test type: %q", test.Type)
```

Comment:
If our tests are written properly, this code block is unreachable. Instead of establishing `Type` as an arbitrary string value in the test struct, we could store the expected constant(s) (ex. `SENSU_E2E_HANDLE_RESPONSE`) there to minimize the # of lines of code.

Reply:
Can you show me what this might look like? How will we know what to compare against `out`?

Reply:
I believe this would eliminate the need for the switch cases and unreachable code. Note: I did not attempt to run this example haha.

```
 	tests := []struct {
  		Name		string
  		Response	string
  		Error		string
  		Env		map[string]string
  		Event		*types.Event
  		ExpErr		bool
  	}{
  		{
  			Name:  "no error",
  			Response: SENSU_E2E_HANDLE_RESPONSE,
  			Error: SENSU_E2E_HANDLE_ERROR,
  			Event: handleEvt,
  			Env: map[string]string{
  				"SENSU_E2E_HANDLE_RESPONSE": "{}",
  			},
  		}, // ... other test cases here
  	}
  
  	for _, test := range tests {
  		t.Run(fmt.Sprintf("%s %s", test.Type, test.Name), func(t *testing.T) {
  			cleanup := newExtension(t, ports[0], test.Env)
  
  			body, _ := json.Marshal(test.Event)
  			resp, err := client.R().SetBody(body).Put(
  				fmt.Sprintf("/events/%s/%s", test.Event.Entity.ID, test.Event.Check.Name))
  			if err != nil {
  				t.Fatal(err)
  			}
  			if status := resp.StatusCode(); status >= 400 {
  				t.Fatalf("bad status: %d", status)
  			}
  
  			out, err := cleanup()
  			if err != nil && !test.ExpErr {
  				fmt.Println(out)
  				t.Fatal(err)
  			}

  			if err == nil {
  					assert.Equal(t, test.Env[test.Response], out)
  				} else {
  					assert.Equal(t, test.Env[test.Error], out)
  				}
  			}
  	}
```